pkg/ingester/ingester.go

```diff
@@ -1673,6 +1747,45 @@ func (i *Ingester) createTSDB(userID string) (*userTSDB, error) {
 	}
 
 	i.tsdbMetrics.setRegistryForUser(userID, tsdbPromReg)
+
+	userDB.ephemeralSeriesRetentionPeriod = i.cfg.BlocksStorageConfig.EphemeralTSDB.Retention
+	userDB.ephemeralFactory = func() (*tsdb.Head, error) {
+		// TODO: check user limit for ephemeral series. If it's 0, don't create head and return error.
+
+		headOptions := &tsdb.HeadOptions{
+			ChunkRange:                     i.cfg.BlocksStorageConfig.EphemeralTSDB.Retention.Milliseconds(),
+			ChunkDirRoot:                   filepath.Join(udir, "ephemeral_chunks"),
+			ChunkPool:                      nil,
+			ChunkWriteBufferSize:           i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadChunksWriteBufferSize,
+			ChunkEndTimeVariance:           i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadChunksEndTimeVariance,
+			ChunkWriteQueueSize:            i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadChunksWriteQueueSize,
+			StripeSize:                     i.cfg.BlocksStorageConfig.EphemeralTSDB.StripeSize,
+			SeriesCallback:                 nil, // TODO: handle limits.
+			EnableExemplarStorage:          false,
+			EnableMemorySnapshotOnShutdown: false,
+			IsolationDisabled:              true,
+			PostingsForMatchersCacheTTL:    i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadPostingsForMatchersCacheTTL,
+			PostingsForMatchersCacheSize:   i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadPostingsForMatchersCacheSize,
+			PostingsForMatchersCacheForce:  i.cfg.BlocksStorageConfig.EphemeralTSDB.HeadPostingsForMatchersCacheForce,
+		}
+
+		headOptions.MaxExemplars.Store(0)
+		headOptions.OutOfOrderTimeWindow.Store(0)
+		headOptions.OutOfOrderCapMax.Store(int64(tsdb.DefaultOutOfOrderCapMax)) // We need to set this, despite OOO time window being 0.
+		headOptions.EnableNativeHistograms.Store(false)
+
+		h, err := tsdb.NewHead(prometheus.WrapRegistererWithPrefix(ephemeralPrometheusMetricsPrefix, tsdbPromReg), log.With(userLogger, storageKey, "ephemeral"), nil, nil, headOptions, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		i.metrics.memEphemeralUsers.Inc()
+
+		// Don't allow ingestion of old samples into ephemeral storage.
+		h.SetMinValidTime(time.Now().Add(-i.cfg.BlocksStorageConfig.EphemeralTSDB.Retention).UnixMilli())
```

Comment:
[non blocking, but I would like to talk more about this] Is this actually correct? What if a customer sends samples with an older timestamp (e.g. while catching up after an outage)? Also we should be consistent with any check done in Push(): we allow any timestamp there (up to TSDB out-of-bounds limits).

Reply:
> What if a customer sends samples with an older timestamp (e.g. while catching up after an outage)?

Yes, that's unfortunately how "ephemeral" storage will work.

> Also we should be consistent with any check done in Push(): we allow any timestamp there (up to TSDB out-of-bounds limits).

We are consistent, because we're updating `minValidTime` of ephemeral storage periodically via `*userDB.Compact` where we call `Truncate` on ephemeral storage.

Reply:
👍